### PR TITLE
PS-9174 Backport bug fixes from MySQL 8.0.37

### PR DESCRIPTION
**Pull request by VarunNagaraju**
https://perconadev.atlassian.net/browse/PS-9174

- Fixes a security issue in mysqldump utility.
- Fixes an assertion in /mysql-8.0.34/sql/field.cc:7119.
- Fixes a failure in Protocol_classic::send_field_metadata.
- Fixes Incorrect results issue when using group by loose index scan.
- Fixes Signal 11 seen in Gtid_set::~Gtid_set.
- Fixes a transaction hang in InnoDB.
- Fixes assertion failure:ibuf0ibuf.cc:3833:ib::fatal.
- Fixes inconsistent FTS state in concurrent scenarios.
- Unifies behaviour when calling plugin->deinit for all plugins.
- Fixes data inconsistencies when high rate of pages split/merge.
- Fixes GTID info lost issue when MySQL crash in Gtid_state::save.
- Post-push fix for MTR test failure percona_statement_set.test.

### CODE REVIEW

File: sql/rpl_gtid_state.cc
Comment (venkatesh-prasad-v) [inline sql/rpl_gtid_state.cc]: Can we rename this to something like below so that test case is more readable?
```suggestion
  DBUG_EXECUTE_IF("crash_before_gtid_persist", DBUG_SUICIDE(););
```

File: mysql-test/suite/binlog/t/binlog_gtid_persistent.test
Comment (venkatesh-prasad-v) [inline mysql-test/suite/binlog/t/binlog_gtid_persistent.test]: I would prefer having
```suggestion
--echo Start the server again to see if server recovers GTIDs from binlogs and updates @@global.gtid_executed and mysql.gtid_executed table.
```

File: mysql-test/suite/binlog/t/binlog_gtid_persistent.test
Comment (venkatesh-prasad-v) [inline mysql-test/suite/binlog/t/binlog_gtid_persistent.test]: `GTID not persisted and recovered` reads as if the recovery was successful but it was not persisted. This is not what happens when we have empty prev_gtids. Can we change it to something like
```suggestion
    --die "Could not find Previous_gtid_log_event in the last binary log parsed by the server"
```

File: mysql-test/suite/binlog/t/binlog_gtid_persistent.test
Comment (venkatesh-prasad-v) [inline mysql-test/suite/binlog/t/binlog_gtid_persistent.test]: `Old GTID` is ambiguous. Can we have something like
```suggestion
    die "Server could not recover GTIDs from binary logs during recovery";
```

**Review by venkatesh-prasad-v — CHANGES_REQUESTED** [review]

File: mysql-test/suite/binlog/t/binlog_gtid_persistent.test
Comment (VarunNagaraju) [reply mysql-test/suite/binlog/t/binlog_gtid_persistent.test]: Done.

File: mysql-test/suite/binlog/t/binlog_gtid_persistent.test
Comment (VarunNagaraju) [reply mysql-test/suite/binlog/t/binlog_gtid_persistent.test]: Done.

File: mysql-test/suite/binlog/t/binlog_gtid_persistent.test
Comment (VarunNagaraju) [reply mysql-test/suite/binlog/t/binlog_gtid_persistent.test]: Done.

File: sql/rpl_gtid_state.cc
Comment (VarunNagaraju) [reply sql/rpl_gtid_state.cc]: Done.

File: mysql-test/suite/binlog/t/binlog_gtid_recovery.test
Comment (venkatesh-prasad-v) [inline mysql-test/suite/binlog/t/binlog_gtid_recovery.test:4]: Please source `include/have_debug.inc` to make it skip when run with release builds.

Comment (VarunNagaraju) [reply mysql-test/suite/binlog/t/binlog_gtid_recovery.test:4]: Done.

**Review by venkatesh-prasad-v — APPROVED** [review]
I've reviewed replication bugfixes and they look good to me. Please wait for approval from @percona-ysorokin and @dlenev for the final approval.

File: mysql-test/suite/binlog/t/binlog_gtid_recovery.test
Comment (percona-ysorokin) [inline mysql-test/suite/binlog/t/binlog_gtid_recovery.test]: Debugging leftovers

File: mysql-test/suite/binlog/r/binlog_gtid_recovery.result
Comment (percona-ysorokin) [inline mysql-test/suite/binlog/r/binlog_gtid_recovery.result:18]: @VarunNagaraju Could you also check if this test passes for the second time when we run with `--repeat=2`.
I see that we have binlog file names (`000001 ` / `000002` / `000003`) recorded here and want to double check.

File: mysql-test/suite/binlog/t/binlog_gtid_recovery.test
Comment (percona-ysorokin) [inline mysql-test/suite/binlog/t/binlog_gtid_recovery.test:17]: Any chance there is a mnemonic for `137`?

File: storage/innobase/ibuf/ibuf0ibuf.cc
Comment (percona-ysorokin) [inline storage/innobase/ibuf/ibuf0ibuf.cc]: Formatting (spaces instead of tabs)

**Review by percona-ysorokin — APPROVED** [review]
LGTM
(with some stylistic comments)

File: mysql-test/suite/binlog/r/binlog_gtid_recovery.result
Comment (venkatesh-prasad-v) [reply mysql-test/suite/binlog/r/binlog_gtid_recovery.result:18]: @percona-ysorokin The test performs `RESET MASTER` in the beginning and it should handle the hardcoded binlog names.

Comment (percona-ysorokin) [reply mysql-test/suite/binlog/r/binlog_gtid_recovery.result:18]: Oops, missed that part somehow. Thanks, Venki. Please, ignore.

File: mysql-test/suite/binlog/t/binlog_gtid_recovery.test
Comment (VarunNagaraju) [reply mysql-test/suite/binlog/t/binlog_gtid_recovery.test]: Removed.

File: mysql-test/suite/binlog/r/binlog_gtid_recovery.result
Comment (VarunNagaraju) [reply mysql-test/suite/binlog/r/binlog_gtid_recovery.result:18]: Yes, tested with **--parallel=10 --repeat=20**

File: mysql-test/suite/binlog/t/binlog_gtid_recovery.test
Comment (VarunNagaraju) [reply mysql-test/suite/binlog/t/binlog_gtid_recovery.test:17]: I couldn't find one.

File: storage/innobase/ibuf/ibuf0ibuf.cc
Comment (VarunNagaraju) [reply storage/innobase/ibuf/ibuf0ibuf.cc]: Fixed.

**Comment by VarunNagaraju** [comment]
https://ps57.cd.percona.com/job/percona-server-5.7-pipeline/37338/

File: mysql-test/r/select_distinct_debug.result
Comment (dlenev) [inline mysql-test/r/select_distinct_debug.result]: Since we have to change test case from 8.0 anyway, perhaps it makes sense to do one more step and move this SET SESSION DEBUG right before SELECT DISTINCT query to make it clear that it should and affects only SELECT DISTINCT statement. What do you think?

**Review by dlenev — COMMENTED** [review]
Hello Venki!

I have only one comment about backport of  PS-9174 "Incorrect results when using group by loose index scan" fix. Please find it below.

I am still looking at two other Optimizer-related changes.

File: mysql-test/r/select_distinct_debug.result
Comment (VarunNagaraju) [reply mysql-test/r/select_distinct_debug.result]: @dlenev Yes, It makes sense. Will update it.

Comment (VarunNagaraju) [reply mysql-test/r/select_distinct_debug.result]: Done.

**Review by dlenev — COMMENTED** [review]
Hello Varun!
It is a pity that we were not able to come up with a test case for PS-9174 "Assertion Failure in /mysql-8.0.34/sql/field.cc:7119" but patch makes sense and seems to be safe to me. I am still looking at the only remaining Optimizer/Executor related change.

**Review by dlenev — COMMENTED** [review]
Hello Varun!

Regarding PS-9174 "Failure in Protocol_classic::send_field_metadata"
I've spent some time trying to produce test case for this issue, but was not successful.
And unlike for another two Optimizer fixes I am still not sure if it applies to 5.7. However, it should be safe to push this change at very least.

**Review by percona-ysorokin — APPROVED** [review]
LGTM

**Comment by VarunNagaraju** [comment]
Complete testrun : https://ps57.cd.percona.com/job/percona-server-5.7-param/610/
ASAN testrun : https://ps57.cd.percona.com/job/percona-server-5.7-pipeline/37572/
https://perconadev.atlassian.net/browse/PS-9248 has been created to address the ASAN test failures.

**Pull request merged**
